File: testbed_ruby/test/associations/one_to_n_test.rb
require 'test/unit'
require './src-gen-umple/student_l'
require './src-gen-umple/mentor_l'
require './src-gen-umple/program_l'

module CruiseAssociations
class OneToNTest < Test::Unit::TestCase

  def test_cannotCreateNullStudent
    assert_raise RuntimeError do #"Unable to create student due to mentor"
      StudentL.new(99,nil)
    end
  end
  
  def test_CreateMentorWithoutStudent
    m = MentorL.new("blah")
    assert_equal(0,m.number_of_students)
  end
  
  def test_CannotSetMentorNull
    m = MentorL.new("blah")
    s = StudentL.new(99,m)
    
    assert_equal(false,s.set_mentor(nil))
  end  
  
  def test_CannotSetMentorAlreadyAtMax
    m = MentorL.new("blah")
    m2 = MentorL.new("blah2")
    
    StudentL.new(99,m)
    StudentL.new(98,m)    
    
    StudentL.new(97,m2)
    s = StudentL.new(96,m2)
    assert_equal(false,s.set_mentor(m))
    
    assert_equal(2,m.number_of_students)
    assert_equal(2,m2.number_of_students)
    assert_equal(m2,s.get_mentor)
  end    
  
  def test_CannotSetMentorExistingAtMin
    m = MentorL.new("blah")
    m2 = MentorL.new("blah2")
    
    StudentL.new(99,m)
    StudentL.new(98,m)
    
    StudentL.new(97,m2)
    s = StudentL.new(96,m2)
    assert_equal(false,s.set_mentor(m))
    
    assert_equal(2,m.number_of_students)
    assert_equal(2,m2.number_of_students)
    assert_equal(m2,s.get_mentor)
  end   
  
  def test_CreateStudentFromMentor
    m = MentorL.new("blah")
    s = StudentL.new(99,m)
  
    assert_equal(99,s.get_number)
    assert_equal(m,s.get_mentor)
    assert_equal(1,m.number_of_students)
    assert_equal(s,m.get_student(0))
  end  
  
  def test_add_studentConstructorInformation
    m = MentorL.new("blah")
    s = StudentL.new(10,m)
  
    assert_equal(10,s.get_number)
    assert_equal(m,s.get_mentor)
    assert_equal(1,m.number_of_students)
    assert_equal(s,m.get_student(0))
  end
  
  def test_add_studentConstructorInformation_tooMany
    m = MentorL.new("blah")
    s = StudentL.new(10,m)
    s = StudentL.new(11,m)

    assert_raise RuntimeError do #"Unable to create student due to mentor"
      s = StudentL.new(13,m)
    end
  end
  
  def test_createStudentWhenMentorAlreadyHasEnogh
    assert_raise RuntimeError do #"Unable to create student due to mentor"
      m = MentorL.new("blah")
      StudentL.new(10,m)
      StudentL.new(11,m)
      StudentL.new(12,m)
      StudentL.new(13,m)
    end
  end  
  
  def test_cannotReplaceMentor
    m = MentorL.new("blah")
    m2 = MentorL.new("blah2")
  
    s = StudentL.new(123,m)
    s2 = StudentL.new(125,m2)
    
    assert_equal(false,s.set_mentor(m2))
    assert_equal(m,s.get_mentor)
    assert_equal(1,m2.number_of_students)
    
    assert_equal(1,m.number_of_students)
  end 
  
  def test_cannotReassign
    m = MentorL.new("blah")
    m2 = MentorL.new("blah2")
  
    s = StudentL.new(123,m)
    assert_equal(false,m2.add_student(s))
  end  
  
  def test_cannotRemoveFromExistingMentor
    m = MentorL.new("blah")
    m2 = MentorL.new("blah2")
  
    s1 = StudentL.new(123,m)
    StudentL.new(124,m)
    
    assert_equal(2,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    
    assert_equal(false,m2.add_student(s1))
  end
  
  def test_is_number_of_students_valid
    m = MentorL.new("blah")
    assert_equal(false,m.is_number_of_students_valid)
  
    StudentL.new(99,m)
    assert_equal(false,m.is_number_of_students_valid)
  
    StudentL.new(98,m)
    assert_equal(true,m.is_number_of_students_valid)
  end
  
  def test_getBoundsForStudent
    assert_equal(2,MentorL::minimum_number_of_students)
    assert_equal(2,MentorL::maximum_number_of_students)
  end  
  
  def test_add_studentWhenMentorHasTooMany
    m = MentorL.new("blah")
    m2 = MentorL.new("blah2")
    
    s = StudentL.new(12,m)
    StudentL.new(13,m)
    
    StudentL.new(21,m2)
    StudentL.new(22,m2)
    
    assert_equal(false,m.add_student(s))
  
    assert_equal(2,m.number_of_students)
    assert_equal(2,m2.number_of_students)
  end  
  
  def test_add_studentWhenMentorHasTooFew
    m = MentorL.new("blah")
    m2 = MentorL.new("blah2")
    
    StudentL.new(12,m)
    StudentL.new(13,m)
    
    StudentL.new(21,m2)
    s = StudentL.new(23,m2)
    
    assert_equal(false,m.add_student(s))
  
    assert_equal(2,m.number_of_students)
    assert_equal(2,m2.number_of_students)
   end  
end
end

File: testbed_php/test/associations/OneToNTest.php
<?php

class OneToNTest extends UnitTestCase
{

  public function test_cannotCreateNullStudent()
  {
  	$this->expectException(new Exception("Unable to create student due to mentor"));
    new StudentL(99,null);
  }
  

  public function test_CreateMentorWithoutStudent()
  {
    $m = new MentorL("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }
  

  public function test_CannotSetMentorNull()
  {
    $m = new MentorL("blah");
    $s = new StudentL(99,$m);
    
    $this->assertEqual(false,$s->setMentor(null));
  }  


  public function test_CannotSetMentorAlreadyAtMax()
  {
    $m = new MentorL("blah");
    $m2 = new MentorL("blah2");
    
    new StudentL(99,$m);
    new StudentL(98,$m);
    
    new StudentL(97,$m2);
    $s = new StudentL(96,$m2);
    $this->assertEqual(false,$s->setMentor($m));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(2,$m2->numberOfStudents());
    $this->assertEqual($m2,$s->getMentor());
  }    


  public function test_CannotSetMentorExistingAtMin()
  {
    $m = new MentorL("blah");
    $m2 = new MentorL("blah2");
    
    new StudentL(99,$m);
    new StudentL(98,$m);
    
    new StudentL(97,$m2);
    $s = new StudentL(96,$m2);
    $this->assertEqual(false,$s->setMentor($m));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(2,$m2->numberOfStudents());
    $this->assertEqual($m2,$s->getMentor());

  
  }   
  

  public function test_CreateStudentFromMentor()
  {
    $m = new MentorL("blah");
    $s = new StudentL(99,$m);

    $this->assertEqual(99,$s->getNumber());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(1,$m->numberOfStudents());
    $this->assertEqual($s,$m->getStudent_index(0));
  }  
  

  public function test_addStudentViaConstructorInformation()
  {
    $m = new MentorL("blah");
    $s = $m->addStudentVia(10);

    $this->assertEqual(10,$s->getNumber());
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(1,$m->numberOfStudents());
    $this->assertEqual($s,$m->getStudent_index(0));
  }


  public function test_addStudentViaConstructorInformation_tooMany()
  {
    $m = new MentorL("blah");
    $m->addStudentVia(10);
    $m->addStudentVia(11);
    $m->addStudentVia(12);
    $this->assertEqual(null,$m->addStudentVia(13));
  }
  
  public function test_createStudentWhenMentorAlreadyHasEnogh()
  {
  	$this->expectException(new Exception("Unable to create student due to mentor"));
    $m = new MentorL("blah");
    new StudentL(10,$m);
    new StudentL(11,$m);
    new StudentL(12,$m);
    new StudentL(13,$m);
  }  
  
  

  public function test_cannotReplaceMentor()
  {
    $m = new MentorL("blah");
    $m2 = new MentorL("blah2");

    $s = $m->addStudentVia(123);
    $m2->addStudentVia(125);
    
    $this->assertEqual(false,$s->setMentor($m2));
    $this->assertEqual($m,$s->getMentor());
    $this->assertEqual(1,$m2->numberOfStudents());
    
    $this->assertEqual(1,$m->numberOfStudents());
    
  } 
  

  public function test_cannotReassign()
  {
    $m = new MentorL("blah");
    $m2 = new MentorL("blah2");

    $s = $m->addStudentVia(123);
    $this->assertEqual(false,$m2->addStudent($s));
  }  
  

  public function test_cannotRemoveFromExistingMentor()
  {
    $m = new MentorL("blah");
    $m2 = new MentorL("blah2");

    $s1 = $m->addStudentVia(123);
    $m->addStudentVia(124);
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(0,$m2->numberOfStudents());
    
    $this->assertEqual(false,$m2->addStudent($s1));
  }
  

  public function test_isNumberOfStudentsValid()
  {
    $m = new MentorL("blah");
    $this->assertEqual(false,$m->isNumberOfStudentsValid());

    new StudentL(99,$m);
    $this->assertEqual(false,$m->isNumberOfStudentsValid());

    new StudentL(98,$m);
    $this->assertEqual(true,$m->isNumberOfStudentsValid());

  }
  

  public function test_getBoundsForStudent()
  {
    $this->assertEqual(2,MentorL::minimumNumberOfStudents());
    $this->assertEqual(2,MentorL::maximumNumberOfStudents());
  }  


  public function test_addStudentWhenMentorHasTooMany()
  {
    $m = new MentorL("blah");
    $m2 = new MentorL("blah2");
    
    $m->addStudentVia(12);
    $m->addStudentVia(13);
    
    $m2->addStudentVia(21);
    $m2->addStudentVia(22);
    $s = $m2->addStudentVia(23);
    
    $this->assertEqual(false,$m->addStudent($s));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(2,$m2->numberOfStudents());
  }  


  public function test_addStudentWhenMentorHasTooFew()
  {
    $m = new MentorL("blah");
    $m2 = new MentorL("blah2");
    
    $m->addStudentVia(12);
    $m->addStudentVia(13);
    
    $m2->addStudentVia(21);
    $s = $m2->addStudentVia(23);
    
    $this->assertEqual(false,$m->addStudent($s));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(2,$m2->numberOfStudents());
  }  
}
